Potential crash when dealing with out of bounds range requests for Blob URLs

cdumez opened the pull request:
#### cc89df4bf0986b6bdb5466a35bfd37a47a60add6
<pre>
Potential crash when dealing with out of bounds range requests for Blob URLs
<a href="https://bugs.webkit.org/show_bug.cgi?id=300867">https://bugs.webkit.org/show_bug.cgi?id=300867</a>
<a href="https://rdar.apple.com/161573171">rdar://161573171</a>

Reviewed by Darin Adler.

The logic dealing with range requests for blob URLs did not validate
the bounds sufficiently to make sure they are valid for the given blob
size. As a result, we could hit a release assertion inside
`std::span::subspan()`.

After this PR, we properly validate the bounds of the range request. If
they cannot be satisfied, we ignore the range request and we return the
full content, as we are allowed to do. This seems to match Chrome&apos;s
behavior also.

Also fix a bug where range requests were not supported with Blob URLs in
WebKitLegacy when using synchronous XMLHttpRequest. This bug was found
during testing.

I also moved some of the code from BlobResourceHandle and NetworkDataTaskBlob
to a new BlobResourceHandleBase common base class, to promote more code
sharing. I will follow up to move even more code to this base class and
reduce duplication further.

Tests:
    fast/files/blob-range-request.html
    fast/files/blob-range-request-sync.html

* LayoutTests/fast/files/blob-range-request-expected.txt: Added.
* LayoutTests/fast/files/blob-range-request-sync-expected.txt: Added.
* LayoutTests/fast/files/blob-range-request-sync.html: Added.
* LayoutTests/fast/files/blob-range-request.html: Added.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/xhr/blob-range.any-expected.txt: Removed.
* Source/WebCore/platform/network/BlobResourceHandle.cpp:
(WebCore::BlobResourceHandleBase::BlobResourceHandleBase):
(WebCore::BlobResourceHandleBase::adjustAndValidateRangeBounds):
(WebCore::BlobResourceHandle::BlobResourceHandle):
(WebCore::BlobResourceHandle::doStart):
(WebCore::BlobResourceHandle::didGetSize):
(WebCore::BlobResourceHandleBase::seek):
(WebCore::BlobResourceHandle::readSync):
(WebCore::BlobResourceHandle::readDataSync):
(WebCore::BlobResourceHandle::readFileSync):
(WebCore::BlobResourceHandle::readDataAsync):
(WebCore::BlobResourceHandle::readFileAsync):
(WebCore::BlobResourceHandle::notifyResponseOnSuccess):
(WebCore::BlobResourceHandle::seek):
* Source/WebCore/platform/network/BlobResourceHandle.h:
Fix the issue for WebKit1.

* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::parseRange):
(WebCore::isCrossOriginSafeRequestHeader):
* Source/WebCore/platform/network/HTTPParsers.h:
Modernize parseRange() to use std::optional instead of integers with
special values when they&apos;re missing.

* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::NetworkDataTaskBlob::resume):
(WebKit::NetworkDataTaskBlob::didGetSize):
(WebKit::NetworkDataTaskBlob::dispatchDidReceiveResponse):
(WebKit::NetworkDataTaskBlob::readData):
(WebKit::NetworkDataTaskBlob::readFile):
(WebKit::NetworkDataTaskBlob::seek):
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.h:
Fix the issue for WebKit2.

Canonical link: <a href="https://commits.webkit.org/301717@main">https://commits.webkit.org/301717@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3490fb7766177ecb58247903d7acdbc71fbf6e41

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126807 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46448 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37423 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133755 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78422 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bbd8a12b-0860-4951-9d6a-f6f7873f7b82) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128678 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47074 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54984 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96494 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f850554a-6521-4890-a828-7a4191986f70) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129755 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37668 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113427 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77014 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/aa7024c4-01f0-4482-b180-be7b34319bd6) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36544 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31604 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77149 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/107498 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31909 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136335 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53480 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41164 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105019 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53976 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109786 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104711 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26706 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50209 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28552 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50943 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53409 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59211 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52665 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55999 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54413 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on LayoutTests/fast/files/blob-range-request.html:
```diff
@@ -0,0 +1,83 @@
+
+function finish() {
+    URL.revokeObjectURL(blobURL);
+     URL.revokeObjectURL(emptyBlobURL);
```
```suggestion
    URL.revokeObjectURL(emptyBlobURL);
```

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.cpp:
```diff
@@ -220,10 +220,40 @@ void BlobResourceHandle::doStart()
+        // m_rangeEnd indicates the last bytes to read.
+        if (*m_rangeEnd > m_totalSize) {
+            m_rangeStart = 0;
+            m_rangeEnd = m_totalSize ? (m_totalSize - 1) : 0;
```
Why the -1?

I know it's not new, just moved code. But mystified; is the range end the offset _before_ the last byte?

cdumez replied:
Those are indexes. For example, if the clients want to fetch the first 1024 bytes, it should use this range header value: `bytes=0-1023`. It's zero-based. In this example, `m_rangeStart` is `0` and `m_rangeEnd` is `1023`. So I believe it is correct to set `m_rangeEnd` to `m_totalSize-1` when we return the rest of the data (till the end of the blob).

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.cpp:
```diff
@@ -277,15 +307,15 @@ void BlobResourceHandle::didGetSize(long long size)
     // The size passed back is the size of the whole file. If the underlying item is a sliced file, we need to use the slice length.
     const BlobDataItem& item = m_blobData->items().at(m_sizeItemCount);
-    size = item.length();
+    size_t updatedSize = static_cast<size_t>(item.length());
```
I don't fully understand why this function now completely ignores its `size` argument. Should we remove the argument if it's not needed?

I'm unsure why the cast to `size_t` guaranteed to work; on a 32-bit platform could it be bigger than size_t if the file is bigger than 4GB? Maybe there's some checking elsewhere that guarantees that?

cdumez replied:
At the end of the day, this ends up with a span and calling `subspan()` which works with `size_t` so I figured I'd use `size_t` everywhere, for consistency. I guess we could use uint64_t here and we'd still end up truncating to a `size_t` in this end when calling `subspan()` AFAICT.

```
NetworkDataTaskBlob::readData(const BlobDataItem& item)
{
// ...
    RefPtr data = item.data();
    auto dataSpan = data->span().subspan(item.offset() + m_currentItemReadSize, static_cast<size_t>(bytesToRead));
```

which ends up calling:
```
class DataSegment : public ThreadSafeRefCounted<DataSegment> {
public:
    size_t size() const { return span().size(); }
    WEBCORE_EXPORT std::span<const uint8_t> span() const LIFETIME_BOUND;
```

Seems like I am not making the situation worse, do you agree?

darinadler replied:
Regarding the change in the casting, I agree that you made it no worse since we’ll eventually get to `size_t`. But I still think we should think about whether there’s a 32-bit bug here due to casting `long long` to `size_t` without any checking.

I didn’t see your thought on my other comment:
> Should we remove the argument if it's not needed?

cdumez replied:
This argument is used in this function though?
```
if (size == -1) {
        notifyFail(Error::NotFoundError);
        return;
    }
```
While we know the size of the file and therefore we theoretically don't need the parameter, we still ask the file reader for the size in case it has moved or changed, in which case the parameter would be -1.

darinadler commented on Source/WebCore/platform/network/HTTPParsers.h:
```diff
@@ -97,8 +97,8 @@ AtomString extractReasonPhraseFromHTTPStatusLine(const String&);
 
-// -1 could be set to one of the return parameters to indicate the value is not specified.
-WEBCORE_EXPORT bool parseRange(StringView, RangeAllowWhitespace, long long& rangeStart, long long& rangeEnd);
+// std::nullopt could be set to one of the return parameters to indicate the value is not specified.
```
```suggestion
// Either of the out arguments could be set to std::nullopt to indicate the value is not specified.
```

darinadler commented on Source/WebCore/platform/network/HTTPParsers.h:
```diff
@@ -97,8 +97,8 @@ AtomString extractReasonPhraseFromHTTPStatusLine(const String&);
-// -1 could be set to one of the return parameters to indicate the value is not specified.
-WEBCORE_EXPORT bool parseRange(StringView, RangeAllowWhitespace, long long& rangeStart, long long& rangeEnd);
+// std::nullopt could be set to one of the return parameters to indicate the value is not specified.
+WEBCORE_EXPORT bool parseRange(StringView, RangeAllowWhitespace, std::optional<size_t>& rangeStart, std::optional<size_t>& rangeEnd);
```
Should we change this to use a return value instead of out arguments?

cdumez replied:
Yes, since I'm modernizing this function, I might as well. I'll make this change before landing.

cdumez replied:
Planning to use:
```
// std::nullopt could be set to one of the range bounds.
struct HTTPRange {
    std::optional<uint64_t> start;
    std::optional<uint64_t> end;
};
WEBCORE_EXPORT std::optional<HTTPRange> parseRange(StringView, RangeAllowWhitespace);
```

darinadler commented on Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
```diff
@@ -142,6 +142,32 @@ void NetworkDataTaskBlob::resume()
     });
 }
 
+auto NetworkDataTaskBlob::adjustAndValidateRangeBounds() -> std::optional<Error>
```
Could we refactor so we share one copy of this code with `BlobResourceHandle`?

I'm mildly shocked at how much of this code looks like it's copied and pasted between these two classes.

cdumez replied:
Yes, I was planning to do this in a follow-up if you don't mind.

cdumez replied:
In this latest iteration, I introduced a new `BlobResourceHandleBase` base class to avoid adding new code duplication in this PR at least. In a follow-up, I will move more of the existing logic from the subclasses to `BlobResourceHandleBase`.

darinadler replied:
Sounds great!

darinadler commented on Source/WebCore/platform/network/BlobResourceHandle.cpp:
```diff
@@ -379,14 +403,14 @@ int BlobResourceHandle::readDataSync(const BlobDataItem& item, std::span<uint8_t
 
-    long long remaining = item.length() - m_currentItemReadSize;
-    long long bytesToRead = std::min(std::min<long long>(remaining, buffer.size()), m_totalRemainingSize);
+    size_t remaining = item.length() - m_currentItemReadSize;
```
Is this right for 32-bit systems or could it truncate?

cdumez replied:
Same comment as above. BlobDataItem relies on DataSegment which relies on span and thus size_t, so I believe we would end up truncating no matter what?

Although, thinking more about it, a Blob could include several items, each fitting into a size_t but the overall size of the Blob not fitting into a size_t. As a result, I believe you're right and it'd be safer to use `uint64_t` when referring to overall blob sizes and not blob item sizes.

cdumez replied:
I am using uint64_t in the latest iteration.

cdumez committed:
Potential crash when dealing with out of bounds range requests for Blob URLs
https://bugs.webkit.org/show_bug.cgi?id=300867
rdar://161573171

Reviewed by Darin Adler.

The logic dealing with range requests for blob URLs did not validate
the bounds sufficiently to make sure they are valid for the given blob
size. As a result, we could hit a release assertion inside
`std::span::subspan()`.

After this PR, we properly validate the bounds of the range request. If
they cannot be satisfied, we ignore the range request and we return the
full content, as we are allowed to do. This seems to match Chrome's
behavior also.

Also fix a bug where range requests were not supported with Blob URLs in
WebKitLegacy when using synchronous XMLHttpRequest. This bug was found
during testing.

I also moved some of the code from BlobResourceHandle and NetworkDataTaskBlob
to a new BlobResourceHandleBase common base class, to promote more code
sharing. I will follow up to move even more code to this base class and
reduce duplication further.

Tests:
    fast/files/blob-range-request.html
    fast/files/blob-range-request-sync.html

* LayoutTests/fast/files/blob-range-request-expected.txt: Added.
* LayoutTests/fast/files/blob-range-request-sync-expected.txt: Added.
* LayoutTests/fast/files/blob-range-request-sync.html: Added.
* LayoutTests/fast/files/blob-range-request.html: Added.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/xhr/blob-range.any-expected.txt: Removed.
* Source/WebCore/platform/network/BlobResourceHandle.cpp:
(WebCore::BlobResourceHandleBase::BlobResourceHandleBase):
(WebCore::BlobResourceHandleBase::adjustAndValidateRangeBounds):
(WebCore::BlobResourceHandle::BlobResourceHandle):
(WebCore::BlobResourceHandle::doStart):
(WebCore::BlobResourceHandle::didGetSize):
(WebCore::BlobResourceHandleBase::seek):
(WebCore::BlobResourceHandle::readSync):
(WebCore::BlobResourceHandle::readDataSync):
(WebCore::BlobResourceHandle::readFileSync):
(WebCore::BlobResourceHandle::readDataAsync):
(WebCore::BlobResourceHandle::readFileAsync):
(WebCore::BlobResourceHandle::notifyResponseOnSuccess):
(WebCore::BlobResourceHandle::seek):
* Source/WebCore/platform/network/BlobResourceHandle.h:
Fix the issue for WebKit1.

* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::parseRange):
(WebCore::isCrossOriginSafeRequestHeader):
* Source/WebCore/platform/network/HTTPParsers.h:
Modernize parseRange() to use std::optional instead of integers with
special values when they're missing.

* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::NetworkDataTaskBlob::resume):
(WebKit::NetworkDataTaskBlob::didGetSize):
(WebKit::NetworkDataTaskBlob::dispatchDidReceiveResponse):
(WebKit::NetworkDataTaskBlob::readData):
(WebKit::NetworkDataTaskBlob::readFile):
(WebKit::NetworkDataTaskBlob::seek):
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.h:
Fix the issue for WebKit2.

Canonical link: https://commits.webkit.org/301717@main
